### PR TITLE
[Graphbolt] Add compact for node pairs

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Add node pairs compact function, which is used in link prediction sampler used to compact postive and negative graphs.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 1]
```diff
@@ -300,6 +302,83 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_neighbors_for_pairs(
+        self,
+        node_pairs,
+    ):
+        """
+        Sample the neighboring edges of the nodes formed by all connected
+        nodes in the provided node pairs. Return the induced subgraph and the
```
Comment (BarclayII) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:311]:
```
        Sample the neighboring edges of the nodes formed by all connected
        nodes in the provided node pairs
```
I don't quite understand this sentence.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 2]
```diff
@@ -300,6 +302,83 @@ def sample_neighbors(
...
+
+        compacted_nodes_dict = defaultdict(list)
+        for ntype, nodes in nodes_dict.items():
+            collected_nodes = torch.cat(nodes)
+            # Compact and find unique nodes.
+            unique_nodes, collected_nodes = torch.unique(
+                collected_nodes, return_inverse=True
+            )
+            ntype_id = node_type_to_id[ntype]
+            # Convert heterogeneous unqiue node id to homogeneous node id.
```
Comment (BarclayII) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:357]: `unique`

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 3]
```diff
@@ -300,6 +302,83 @@ def sample_neighbors(
...
+        if self.metadata:
+            node_type_to_id = self.metadata.node_type_to_id
+        node_type_offset = self.node_type_offset
+        is_mapping = isinstance(node_pairs, Mapping)
+        # Treat it as a homogeneous graph.
+        if is_mapping:
+            assert (
+                node_type_to_id is not None and node_type_offset is not None
+            ), "Please note that sampling with different edge types is \
+                not supported in a homogeneous graph."
```
Comment (BarclayII) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:336]: Is this not going to be supported, or is it just "not yet implemented"?

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 4]
```diff
@@ -300,6 +302,83 @@ def sample_neighbors(
...
+            u, v = node_pair
+            u_size, v_size = u.numel(), v.numel()
+            u = compacted_nodes_dict[u_type][:u_size]
+            compacted_nodes_dict[u_type] = compacted_nodes_dict[u_type][u_size:]
+            v = compacted_nodes_dict[v_type][:v_size]
+            compacted_nodes_dict[v_type] = compacted_nodes_dict[v_type][v_size:]
+            compacted_node_pairs[etype] = (u, v)
+        if not is_mapping:
+            compacted_node_pairs = list(compacted_node_pairs.values())[0]
+        # Perform negighbor sampling.
```
Comment (BarclayII) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:375]: `neighbor`

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 2 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:357]: fixed

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 4 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:375]: fixed

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 1 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:311]: changed for clarify

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 3 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:336]: It is not supported.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 5]
```diff
@@ -304,6 +306,85 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_neighbors_for_pairs(
+        self,
+        node_pairs,
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:311]: nit: data type.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 6]
```diff
@@ -304,6 +306,85 @@ def sample_neighbors(
             nodes, fanouts.tolist(), replace, return_eids, probs_or_mask
         )
 
+    def sample_neighbors_for_pairs(
```
Comment (frozenbugs) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:309]:
```suggestion
    def sample_neighbors_with_node_pairs(
```

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 7]
```diff
@@ -304,6 +306,85 @@ def sample_neighbors(
...
+            compacted_nodes_dict[u_type] = compacted_nodes_dict[u_type][u_size:]
+            v = compacted_nodes_dict[v_type][:v_size]
+            compacted_nodes_dict[v_type] = compacted_nodes_dict[v_type][v_size:]
+            compacted_node_pairs[etype] = (u, v)
+        if not is_mapping:
+            compacted_node_pairs = list(compacted_node_pairs.values())[0]
+        # Perform neighbor sampling.
+        unique_nodes = torch.cat(unique_nodes_all_types)
+        return (
+            compacted_node_pairs,
```
Comment (czkkkkkk) [inline python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:384]: After compacting the node pairs, what would be the dst nodes when we call `dgl.to_block`?

Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:384]: The dst nodes should be the seed nodes in the sampled subgraph,

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 6 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:309]: changed for fix

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py (hunk 5 above)
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:311]: add description in docstring.

File: python/dgl/graphbolt/graph_storage/csc_sampling_graph.py [hunk 8]
```diff
@@ -304,6 +306,85 @@ def sample_neighbors(
...
+            - If `node_pairs` is a tuple: It should be in the format ('u', 'v')
+            representing source and destination pairs.
+            - If `node_pairs` is a dictionary: The keys should be edge type and
+            the values should be corresponding node pairs.
+        Returns
+        -------
+        Tuple[node_pairs, CSCSamplingGraph]
+            The compacted node pairs and the created CSCSamplingGraph object.
+        """
+
```
[the remark this answers is not in the document]
Comment (peizhou001) [reply python/dgl/graphbolt/graph_storage/csc_sampling_graph.py:330]: removed.
